misc/testerina/modules/testerina-core/src/main/ballerina/assert.bal

```diff
@@ -69,9 +71,8 @@ public isolated function assertFalse(boolean condition, string msg = "Assertion
 # + actual - Actual value
 # + expected - Expected value
 # + msg - Assertion error message
-public isolated function assertEquals(anydata|error actual, anydata|error expected, string msg = "Assertion Failed!") {
-    boolean isEqual = (actual == expected);
-    if (!isEqual) {
+public isolated function assertEquals(Comparable actual, Comparable expected, string msg = "Assertion Failed!") {
```

Comment:
IIRC, `assertEquals` was meant to do the same as `==` and `assertExactEquals` the same as `===`. I think it'll be consistent to make this
```suggestion
public isolated function assertEquals(anydata actual, anydata expected, string msg = "Assertion Failed!") {
```
instead.

Reply:
@MaryamZi I changed this like this to have the same behavior testerina users used to have with this API.

@hevayo please confirm if the suggested is the expected behavior.

Reply:
The reason we didn't include object are because we didn't do a value compare. And the assert equals support printing diff as well. For object equals we use assertExactEquals. is that preserved with this change.

Reply:
@rdhananjaya 

What's the possibility of the below scenario with this change? If the function intAdd returns an error?

`test:assertEquals(4,intAdd(1,3))`

Reply:
Yes, this does not change the diff printing mechanism, only the value comparison.
```ballerina
import ballerina/test;

public function main() {
    error a = error("hello", name = "oh no!!");
    error b = error("hello", name = "oh no!");
    test:assertEquals(a, b);
}
```
Out:
```
error: Assertion Failed!
 
expected: 'error("hello",name="oh no!")'
actual  : 'error("hello",name="oh no!!")'
        at ballerina.test.0_0_0:createBallerinaError(assert.bal:45)
           ballerina.test.0_0_0:assertEquals(assert.bal:77)
           dhananjaya.newProj.0_1_0:main(main.bal:7)
```

If we are to constrain parameters to `anydata`, a programmer who wants to compare two errors will have to write the comparison logic them self, or maybe testerina could supply a separate function to compare errors for similarity.

Reply:
@Dilhasha 
```ballerina
import ballerina/test;

public function main() {
    test:assertEquals(intAdd(1,3), 4);
}

function intAdd(int i, int j) returns int|error {
    return error("opsi");
}
```
Outupt:
```
error: Assertion Failed!
 
expected: <int> '4'
actual  : <error> 'error("opsi")'
        at ballerina.test.0_0_0:createBallerinaError(assert.bal:45)
           ballerina.test.0_0_0:assertEquals(assert.bal:77)
           main:main(main.bal:4)
```

Reply:
@rdhananjaya, agreed about the API change. 

My concern though is that by doing this we are doing a special comparison just for errors, a value equality check that is not defined in the language spec. 

We would then have to explicitly define how we are doing this comparison for `ballerina/test`. 

Additionally, I think users may raise queries as to why this special comparison is only done for errors, and not other behavioural values like objects. One can argue that there can be a similar comparison for objects by comparing field-wise?

IMO, even though it may break code, it would be cleaner to base
- `assertEquals` and `assertNotEquals` on `==` defined in the language
- `assertExactEquals` and `assertNotExactEquals`  on `===` defined in the language

This way we'll have a clear separation in the methods and a rationale based on the language itself.

Yes, the user will have to write their own method to compare errors and structures with errors then, but I think it'll still be consistent with other behavioural values.

Reply:
For me at least, testerina is not a part of the language, it's a test framework just like any other, which helps me execute tests and compare values. 

About objects vs error:
Comparing fields of the object is not sufficient, as there is a behavior associated with each object. I think we don't have a way to compare those.
But for errors, even though they are categorized as behavioral values, I don't see any behavior that we can't compare. And I also think we don't need to compare the stack trace as it renders this comparison pretty useless. 

Note: If we are to go with the current change we need to compare distinct type ids as well.

Reply:
I asked why the error is a behavioral value in the spec: https://github.com/ballerina-platform/ballerina-spec/issues/699

Reply:
Created a discussion to conclude this https://github.com/ballerina-platform/ballerina-lang/discussions/28314
Please provide your opinions.